[Foo Jing Yi] iP

foojingyi opened the pull request.

Commit:
Add Gradle support

foojingyi committed:
Add Message.java with constructor, reply() method, and enum for command types

foojingyi committed:
Added getCmd() method to return command

foojingyi committed:
Created user input feature and respond to user inputs, deleted default output

Commit:
Changed ECHO command type to ADD with appropriate ouput

Commit:
Added a HashSet "list" to keep all added items

Commit:
Added HashSet "list" as variable in Message and to the constructor for Message objects

Commit:
Change HashSet instances to List implemented with ArrayList to preserve order

Commit:
Updated reply() method for ADD command to add to list

Commit:
Change sc.next() to sc.nextLine() so that new Message objects and initialised with the whole of the next line of user input and not with each word of the line as separate objects

Commit:
Added Command type LIST and reply() case for this type

Commit:
Updated chatbot's tone in replies

Commit:
Added Task.java class, adapting from example given

Commit:
Changed list from list of String to list of Task, and added status icons to replies to LIST command

Commit:
Added Command type DONE and updated reply() method for this case

Commit:
Overrode toString method for Task objects such that status icon is part of Task's string representation and not added in Message's reply() method

Commit:
Added Deadline, Event and Todo classes extending Task, adapted from example given

Commit:
Changed accessibility for "private" variables to "protected"

Commit:
Changed ADD Command type to NOT_FOUND for inputs that don't match any commands, added TODO, DEADLINE and EVENT to Command types

Commit:
Changed switch statements to if statements in Message.java, updated reply() for NOT_FOUND command

Commit:
Fixed bug caused by typo in Event class, changed "dl" to "time"

Commit:
Added reply() method for cases of Command types TODO, DEADLINE and EVENT

Commit:
Updated files in text-ui-test to use the I/O redirection technique to test

Commit:
Added DukeException.java to handle exceptions from the chatbot.

Commit:
Added NoCommandException.java and NoDescriptionException.java to handle exceptions as named in the file names

Commit:
Updated Duke.java and Message.java such that incorrect inputs of 1. inputs with no commands and 2. todo/deadline/event commands with no description, are handled as exceptions

Commit:
Added WrongItemIndexException.java to deal with user input errors from "done" and "delete" commands

Commit:
Changed exceptions caught in main class to specifically DukeExceptions

Commit:
Changed errors arising from using "done" commands to be handled with WrongItemIndexException

Commit:
Added DELETE command type and its corresponding reply() method case

Commit:
Ignore "data/duke.txt" when it is created.

Commit:
Read file "data/duke.txt" from hard disk on start up and pass it to each Message object

Commit:
Change format of toString for Deadline and Event so that it is easier to separate date/time from task description

Commit:
Add FileToListReader.java to convert a file from hard disk to list of Task objects

Commit:
Write to file "data/duke.txt" when changes are made to list of tasks

Commit:
Teach duke to parse date and times inputted as "yyyy-MM-dd HH:mm" and output as "d MMM yyyy h.m a"

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8' and resolve conflicts

Commit:
Change Deadline and Event to work with LocalDateTime objects rather than Strings to represent the deadlines and change format of dates and times as inputs

Commit:
Change FileToListReader.java to Storage.java, and added functionality to update file in hard disk
Have not implemented relevant errors

Commit:
Add Ui.java to deal with interactions with user

Commit:
Abstract content of load() and update() methods to new methods parseToStorage() and parseFromStorage()

Commit:
Ignore tasks.txt created through using the chatbot

Commit:
Update welcome message in Ui.java

Commit:
Add TaskList.java to contain the list of task and methods for manipulating this list

Commit:
Add methods to Task.java and its children to support functionality of Storage and TaskList

Commit:
Separate Message.java functionality into Parser.java and Command.java and its children

Commit:
Remove bottom line from error message

Commit:
Update Duke.java with more OOP structure using the new classes

Commit:
Divide classes into packages and refactored accordingly

Commit:
Added JUnit tests including dummyTest() for Duke and tests for Task methods

Commit:
Ignore Manifest file created when jar is built

Commit:
Add JavaDoc comments to all public classes and non-overridden methods

Commit:
Change indentation and variable names to comply with coding standards

Commit:
Add find command

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Resolve conflicts

Commit:
Simplify finding keywords by removing whitespaces and comparing in lowercase

aaronlwm commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+ * @author Foo Jing Yi
+ */
+public class Duke {
+    /** Storage object used by Duke to load from and write to hard disk */
```
Perhaps you can move the description to the class file itself? To reduce clutter

aaronlwm commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+    public void run() {
+        ui.showWelcomeMsg();
+        boolean isExit = false;
+        while (!isExit) {
```
Maybe instead of declaring isExit, how about while(True)... and break/return when condition is reached?

aaronlwm commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,201 @@
+            throw new DukeException("Cannot recognise type");
+        }
+
+        return taskType +
```
Maybe it's more readable to put it in a single line instead

aaronlwm commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Public constructor.
+     */
+    public Ui() { }
```
Not sure whether you really need this, Java has default constructors

foojingyi replied:
I included this for Javadoc purposes as stated [here](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#defaultconstructors), but I think I could include a longer comment and make sure this is consistent for other classes as well. Thank you for your comments!

aaronlwm reviewed (commented):
Nice comprehensive code. LGTM!

WangZijun97 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,201 @@
+            " | " +
+            status +
+            " | " +
+            taskDescription;
```
Perhaps place the operator as part of the next line for ease of reading (easily shows that the line is part of the previous)?
This applies to other instances e.g. in AddCommand

WangZijun97 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,201 @@
+                String taskName = taskElements[2];
+                LocalDateTime dateTime = LocalDateTime.parse(taskElements[3],
+                    DateTimeFormatter.ofPattern("d MMM yyyy, h.m a"));
+                task = new Event(taskName, dateTime);
```
Perhaps this (and other similar chunks of code) should be broken up into distinct logic units with comments to describe what they do?

WangZijun97 commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,20 @@
+public class DeadlineTest {
+
+    @Test
+    public void testStringConversion() {
```
Should this be in the ```featureUnderTest_testScenario_expectedBehavior()``` format?

WangZijun97 reviewed (commented):
Code in general looks quite clean and well formatted in alignment with coding standards for the most part! Only small nits here and there.

Commit:
Change error handling for development purposes

Commit:
Review formatting and moved CommandType enum to its own file

wang-jun-hao commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+ * The Duke program is an interactive bot that offers commands to help the
+ * user keep track of a mutable list of tasks that can be of 3 types:
+ * <code>Todo</code>, <code>Deadline</code> or <code>Event</code>.
+ * These tasks can also be marked as done and will be saved in the hard disk.
```
I like how you summarised the functionality of Duke concisely 👍

wang-jun-hao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,201 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
```
I like how the import statements are separated based on their association

wang-jun-hao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,201 @@
+            }
+        }
+
+        Scanner sc = null;
```
Would it be better to just declare sc without the null initialisation?

wang-jun-hao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,201 @@
+            taskDescription = task.getTaskName() +
+                " | " +
+                ((Event) task).getDateTime().format(
+                    DateTimeFormatter.ofPattern("d MMM yyyy, h.m a"));
```
Perhaps you can abstract out the common code 
`taskDescription = task.getTaskName();`

wang-jun-hao commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,77 @@
+ * User has to type these commands in each line of input
+ * to command the bot.
+ */
+enum CommandType {
+    LIST,
+    DONE,
+    DELETE,
+    BYE,
+    TODO,
+    DEADLINE,
+    EVENT,
+    FIND
+}
```
Perhaps you can create a separate java file for this enum class since it is used by other classes too (DoneCommand, AddCommand, etc)

wang-jun-hao reviewed (commented):
Good adherence to naming conventions overall. Perhaps can improve on abstracting out common code and leave certain variables uninitialised rather than explicitly initialise to null.

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Add checkstyle files and change checkstyle version

Commit:
Add dependencies for JavaFX in build.gradle

Commit:
Try HelloWorld GUI application

Commit:
Add according to JavaFX tutorial 2

Commit:
Add according to JavaFX tutorial 3

Commit:
Add according to JavaFX tutorial 4

Commit:
Convert MainWindow to use the fx:root construct

Commit:
Implement basic GUI

Commit:
Merge branch 'branch-Level-10'

Commit:
Refactor execute command

So that
* can work from CLI and GUI
* reduce repeated code

Commit:
Fix bug where deleting the last item in the list results in an error

Commit:
Add assertions

Commit:
Refactor repeated code in date formatting and task handling

foojingyi committed:
Merge pull request #1 from foojingyi/branch-A-Assertions

Add assertions

Commit:
Merge branch 'master' of https://github.com/foojingyi/ip

Commit:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts: (Resolved)
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java

foojingyi committed:
Merge pull request #2 from foojingyi/branch-A-CodeQuality

Improve Code Quality

Commit:
Merge branch 'master' of https://github.com/foojingyi/ip

Commit:
Improve code quality

* reduce arrowhead code by abstracting methods out
* add new exceptions
* change exception names for readibility

Commit:
Move reply strings to Ui instead of within command objects

Commit:
Specify errors for AddCommand

Commit:
Set up CI

foojingyi committed:
Set up CI from GitHubActions

Commit:
Remove wrongly put .yml file

Commit:
Implement more flexible search

Searches for task using the "Find" command can now include multiple keywords and can tasks can be found even if only matching partially.

Commit:
Create gui package

Commit:
Resolve bug where the first task in a list is added to file with a blank line above

Commit:
Resolve bug where file cannot be loaded

Commit:
Make GUI application quit Exit command

Commit:
Organise import statements for Ui.java

Commit:
Change file on disk to save to.

Commit:
Commit MANIFEST.MF file

Commit:
Rename bot from duke to meimei and changed bot profile picture in GUI

Commit:
Refactor exception classes names and variable names to fit with Meimei

Commit:
Improve coding style according to coding standards

Commit:
Change user profile and changed mentions of duke to meimei

Commit:
Change format of the reader friendly date and times

Commit:
Add Ui.png to docs folder

Commit:
Update README for Meimei Bot

foojingyi committed:
Set theme jekyll-theme-architect

Commit:
Update build.gradle

Commit:
Shorten line in MainWindow.java to fit with coding conventions

Commit:
Delete unnecessary test class and fix import statement order

Commit:
Change to separate imports in TastTest.java

Commit:
Fix files according to date time format
